server/jetstream_api.go

```diff
@@ -1033,6 +1080,193 @@ func (s *Server) jsStreamPurgeRequest(sub *subscription, c *client, subject, rep
+	if !acc.JetStreamEnabled() {
+		resp.Error = jsNotEnabledErr
+		s.sendAPIResponse(c, subject, reply, string(msg), s.jsonResponse(&resp))
+		return
+	}
+	if !isEmptyRequest(msg) {
+		resp.Error = jsNotEmptyRequestErr
+		s.sendAPIResponse(c, subject, reply, string(msg), s.jsonResponse(&resp))
+		return
+	}
+	stream := streamNameFromSubject(subject)
+	if _, err := acc.LookupStream(stream); err == nil {
+		resp.Error = &ApiError{Code: 400, Description: fmt.Sprintf("stream [%q] already exists", stream)}
+		s.sendAPIResponse(c, subject, reply, string(msg), s.jsonResponse(&resp))
+		return
+	}
+
+	tfile, err := ioutil.TempFile("", "jetstream-restore-")
+	if err != nil {
+		resp.Error = &ApiError{Code: 500, Description: "jetstream unable to open temp storage for restore"}
+		s.sendAPIResponse(c, subject, reply, string(msg), s.jsonResponse(&resp))
+		return
+	}
+
+	// Create our internal subscription to accept the snapshot.
+	restoreSubj := fmt.Sprintf("_restore_.%s", nuid.Next())
+
+	// FIXME(dlc) - Can't recover well here if something goes wrong. Could use channels and at least time
+	// things out. Note that this is tied to the requesting client, so if it is a tool this goes away when
+	// the client does. Only thing leaking here is the sub on strange failure.
+	acc.subscribeInternal(c, restoreSubj, func(sub *subscription, c *client, _, _ string, msg []byte) {
+		if len(msg) == 0 {
+			tfile.Seek(0, 0)
+			// TODO(dlc) - no way right now to communicate back.
+			acc.RestoreStream(stream, tfile)
+			tfile.Close()
+			os.Remove(tfile.Name())
+			c.processUnsub(sub.sid)
+			return
+		}
+		// Append chunk to temp file.
+		tfile.Write(msg)
+	})
+
+	resp.DeliverSubject = restoreSubj
+	s.sendAPIResponse(c, subject, reply, string(msg), s.jsonResponse(resp))
+}
+
+// Process a snapshot request.
+func (s *Server) jsStreamSnapshotRequest(sub *subscription, c *client, subject, reply string, msg []byte) {
+	if c.acc == nil {
+		return
+	}
+	acc := c.acc
+
+	var resp = JSApiStreamSnapshotResponse{ApiResponse: ApiResponse{Type: JSApiStreamSnapshotResponseType}}
+	if !acc.JetStreamEnabled() {
+		resp.Error = jsNotEnabledErr
+		s.sendAPIResponse(c, subject, reply, string(msg), s.jsonResponse(&resp))
+		return
+	}
+	if isEmptyRequest(msg) {
+		resp.Error = jsBadRequestErr
+		s.sendAPIResponse(c, subject, reply, string(msg), s.jsonResponse(&resp))
+		return
+	}
+	stream := streamNameFromSubject(subject)
+	mset, err := acc.LookupStream(stream)
+	if err != nil {
+		resp.Error = jsNotFoundError(err)
+		s.sendAPIResponse(c, subject, reply, string(msg), s.jsonResponse(&resp))
+		return
+	}
+
+	var req JSApiStreamSnapshotRequest
+	if err := json.Unmarshal(msg, &req); err != nil {
+		resp.Error = jsInvalidJSONErr
+		s.sendAPIResponse(c, subject, reply, string(msg), s.jsonResponse(&resp))
+		return
+	}
+	if !IsValidSubject(req.DeliverSubject) {
+		resp.Error = jsBadRequestErr
```

Comment:
we really don't want any `jsBadRequestErr` anywhere :) Better to send a `&ApiError{Code: 400, Description: "invalid delivery subject"}` or similar

Reply:
ok